Gradle 5.1+ compatibility

rpalcolea opened the pull request:
Hi @KengoTODA  

This is a fix for https://github.com/spotbugs/spotbugs-gradle-plugin/issues/74.

Because Gradle 5.1 is milestone and RC1 is coming out next week, is it possible for you to release a RC version with this change. Also this is a breaking change so a major version which would be only for 5.1+ is the ideal. 

This is due to changes in internal APIs used by the plugin.

rpalcolea commented:
done, please let me know if the changelog makes sense

KengoTODA commented:
Oh, yes, we need to separate minor release and major one...

Then to make release process simple, I will release minor update in `master` branch first.

KengoTODA commented:
@rpalcolea I've released 1.6.6, so please rebase your topic branch to the latest origin/master.

rpalcolea commented:
Done @KengoTODA ! :)

KengoTODA commented:
build is failed due to:

```
Verification of Gradle distribution failed!
Your Gradle distribution may have been tampered with.
Confirm that the 'distributionSha256Sum' property in your gradle-wrapper.properties file is correct and you are downloading the wrapper from a trusted source.
 Distribution Url: https://services.gradle.org/distributions-snapshots/gradle-5.1-20181207163352+0000-bin.zip
Download Location: /home/travis/.gradle/wrapper/dists/gradle-5.1-20181207163352+0000-bin/3kizo7w5pv931zsz1cmnp4qmx/gradle-5.1-20181207163352+0000-bin.zip
Expected checksum: 'e66e69dce8173dd2004b39ba93586a184628bc6c28461bc771d6835f7f9b0d28'
  Actual checksum: 'e67b3076c968f28954bf6da8d422ecd518bcf644e8cc8b5d2d4c8c9c9bfd3be6'
```

rpalcolea commented:
Updated to latest nightly -> distributions-snapshots/gradle-5.1-20181211114623+0000-bin.zip

I can use this locally

KengoTODA commented:
wmm... it doesn't work in my environment. I tried with both of JDK8 and JDK11.

I'm sure that you need to update `distributionSha256Sum` in gradle-wrapper.properties

rpalcolea commented:
interesting..

jjust upgraded to the latest wrapper using the `wrapper` task 

```
spotbugs-gradle-plugin git/gradle-5-1-support
❯ ./gradlew wrapper --gradle-version 5.1-20181211165224+0000

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed

spotbugs-gradle-plugin git/gradle-5-1-support*
❯ git status
On branch gradle-5-1-support
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   gradle/wrapper/gradle-wrapper.jar
	modified:   gradle/wrapper/gradle-wrapper.properties

no changes added to commit (use "git add" and/or "git commit -a")

spotbugs-gradle-plugin git/gradle-5-1-support*
❯ git diff
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: gradle/wrapper/gradle-wrapper.jar (binary)
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: gradle/wrapper/gradle-wrapper.properties
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ gradle-wrapper.properties:2 @
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.3-bin.zip
distributionUrl=https\://services.gradle.org/distributions-snapshots/gradle-5.1-20181211165224+0000-bin.zip
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
```


It should work for you

rpalcolea commented:
Upgraded to 5.1-rc-1 that was released Today.

Also, removed `append` from the `jacoco` config because it is a deprecated property and it should always be `true`

This is the warning:

```
The append property has been deprecated. This is scheduled to be removed in Gradle 6.0. Append should always be true.
        at jacoco_ankv8eq3sgq8zksl1tsu5sklw$_run_closure3$_closure5.doCall(/spotbugs-gradle-plugin/gradle/jacoco.gradle:15)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
```

KengoTODA commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.1-rc-1-bin.zip
```
To stabilize the build, we can keep previous config:

```suggestion
distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.3-bin.zip
```

and use `.travis.yml` to enable build with Gradle 5.1 RC1.

KengoTODA replied:
In `.travis.yml`, you can add one build matrix item like:

```yml
env:
  matrix:
    - GRADLE_VERSION=default
    - GRADLE_VERSION=5.0
    - GRADLE_VERSION=5.1-rc-1
```

rpalcolea replied:
I don't think this change is backwards compatible, because the API changed, so it would break on Gradle 5.0 or less. This is because `CollectionCallbackActionDecorator` is new in Gradle 5.1 https://github.com/gradle/gradle/commits/master/subprojects/core/src/main/java/org/gradle/api/internal/CollectionCallbackActionDecorator.java

I only kept `    - GRADLE_VERSION=5.1-rc-1` in the matrix for now.

to make this backwards compatible, it would require probably a rewrite on how things are done in the plugin, but I don't think there is a clear guideline from Gradle team on this.

KengoTODA replied:
OK then I believe that what we should do is not merging this PR but finding the way to stop depending on internal API like this `CollectionCallbackActionDecorator`.

rpalcolea commented:
Closing this for now then. The change required is to stop `TaskReportContainer`. Is it possible for you to prioritize this? gradle 5.1 should be out pretty soon

KengoTODA commented:
I tried it, and found that it's quite hard. I'll reopen this PR to merge suggested changes from you.

KengoTODA commented:
Before merge these changes, I need to add several commits into the branch, so I will create another PR that uses the branch in this repo (not in @rpalcolea 's forked repo).

rpalcolea commented:
Hi @KengoTODA , I asked gradle team about this and they probably will rollback or add the old constructor for 5.1 to avoid this breaking change -> https://github.com/gradle/gradle/pull/8076

It would be nice if they provide a public API to prevent this in the future.

rpalcolea commented:
No problem, looks like this breaking change will be part of 5.2 though -> https://github.com/gradle/gradle/pull/8076#issuecomment-449072593
